### DISCUSSION TITLE
Improved Overload Resolution

### DISCUSSION THREAD

**Discussion by 3ok** [microsoft/pylance-release#4148] [Enhancement]
Hi 👋 

I have been using Pylance as my language server for Python development, and I have noticed an issue with overload resolution. Specifically, when I define multiple overloads for a function, Pylance does not always suggest the correct overload based on the arguments provided.

For example, consider the following code:

```python
from typing import Any, Literal, overload, Union

@overload
def foo(name: Literal["bunny"], *, color: str, **kwargs: Any) -> str: ...

@overload
def foo(name: Literal["human"], *, age: int, **kwargs: Any) -> int: ...

@overload
def foo(name: str, **kwargs: Any) -> None: ...

def foo(name: str, **kwargs: Any) -> Union[None, str, int]:
    if (
        name == "bunny" and
        "color" in kwargs and
        isinstance(kwargs["color"], str)
    ):
        return kwargs["color"]
    elif (
        name == "human" and
        "age" in kwargs and
        isinstance(kwargs["age"], int)
    ):
        return kwargs["age"]
    else:
        return None
```

When I start typing a call to the `foo` function with the argument `"human"`, and VSCode/Pylance suggests overloads, it always shows the first overload (the `"bunny"` 🐰 one) , even though it is not the most relevant one in this case.

![image](https://user-images.githubusercontent.com/40218948/227776407-22f383f7-d174-4fc0-9208-9a48c7c07434.png)

I would have expected that either the second or third overload would be shown instead of the first one that is no longer relevant. Automatically going to `2 / 3` or `3 / 3` (or removing the first overload from being shown, i.e. have something like `n / 2`) would be more user-friendly IMO, especially in cases where there is a lot of overloads.

Thanks!

**Comment by erictraut** [comment]
This is related to [this work item](https://github.com/microsoft/pylance-release/issues/3850). We've implemented the necessary support in the type evaluator to determine which overloads are applicable for a call expression. The hover text (tooltip) now applies this filtering, but the signature help (the functionality you're referring to above) has not yet been updated to apply this filtering.

**Reply by 3ok** [reply]
Great to hear that this is being worked on, and I appreciate the update on the current status of the feature. I look forward to seeing the signature help functionality updated to apply the filtering as well. Thank you for the response! 🙂

(I guess we can close this discussion post?)
